File: contests/lc/2147.NumberofWaystoDivideaLongCorridor/solution.py
from typing import List


MOD = 1_000_000_007


def numberOfWays(corridor: str) -> int:
    """Return the number of ways to divide the corridor."""
    seats: List[int] = [i for i, ch in enumerate(corridor) if ch == "S"]

    if len(seats) == 0 or len(seats) % 2 == 1:
        return 0

    ways = 1
    for i in range(2, len(seats), 2):
        gap = seats[i] - seats[i - 1]
        ways = (ways * gap) % MOD

    return ways

File: contests/lc/2147.NumberofWaystoDivideaLongCorridor/solution.go
package numberofwaystodividealongcorridor

// numberOfWays returns the number of ways to split the corridor into sections
// with exactly two seats each. The formula multiplies the gaps between seat
// pairs; see Readme.md for details.
func numberOfWays(corridor string) int {
	const mod = 1_000_000_007

	var seats []int
	for i, c := range corridor {
		if c == 'S' {
			seats = append(seats, i)
		}
	}

	if len(seats) == 0 || len(seats)%2 == 1 {
		return 0
	}

	ans := 1
	for i := 2; i < len(seats); i += 2 {
		gap := seats[i] - seats[i-1]
		ans = (ans * gap) % mod
	}

	return ans
}
